pkg/client/unversioned/remotecommand/remotecommand.go

```diff
@@ -21,141 +21,118 @@ import (
 	"io"
 	"io/ioutil"
 	"net/http"
+	"net/url"
 	"sync"
 
 	"k8s.io/kubernetes/pkg/api"
 	client "k8s.io/kubernetes/pkg/client/unversioned"
-	"k8s.io/kubernetes/pkg/conversion/queryparams"
-	"k8s.io/kubernetes/pkg/runtime"
 	"k8s.io/kubernetes/pkg/util"
 	"k8s.io/kubernetes/pkg/util/httpstream"
 	"k8s.io/kubernetes/pkg/util/httpstream/spdy"
 )
 
-type upgrader interface {
-	upgrade(*client.Request, *client.Config) (httpstream.Connection, error)
+type Executor interface {
+	Stream(stdin io.Reader, stdout, stderr io.Writer, tty bool) error
 }
 
-type defaultUpgrader struct{}
-
-func (u *defaultUpgrader) upgrade(req *client.Request, config *client.Config) (httpstream.Connection, error) {
-	return req.Upgrade(config, spdy.NewRoundTripper)
+type StreamExecutor interface {
+	Executor
+	httpstream.Dialer
 }
 
-type Streamer struct {
-	req    *client.Request
-	config *client.Config
-	stdin  io.Reader
-	stdout io.Writer
-	stderr io.Writer
-	tty    bool
+type streamExecutor struct {
+	upgrader  httpstream.UpgradeRoundTripper
+	transport http.RoundTripper
 
-	upgrader upgrader
+	method string
+	url    *url.URL
 }
 
-// Executor executes a command on a pod container
-type Executor struct {
-	Streamer
-	command []string
-}
+// NewExecutor connects to the provided server and upgrades the connection to
+// multiplexed bidirectional streams. The current implementation uses SPDY,
+// but this could be replaced with HTTP/2 once it's available, or something else.
+// TODO: the common code between this and portforward could be abstracted.
```

Comment:
That's the TODO

On Wed, Oct 7, 2015 at 3:21 PM, Andy Goldstein <notifications@github.com>
wrote:

> In pkg/client/unversioned/remotecommand/remotecommand.go
> <https://github.com/kubernetes/kubernetes/pull/14618#discussion_r41433747>
> :
>
> >  }
> >
> > -// Executor executes a command on a pod container
> > -type Executor struct {
> > -	Streamer
> > -	command []string
> > -}
> > +// NewExecutor connects to the provided server and upgrades the connection to
> > +// multiplexed bidirectional streams. The current implementation uses SPDY,
> > +// but this could be replaced with HTTP/2 once it's available, or something else.
> > +// TODO: the common code between this and portforward could be abstracted.
>
> I think what's bothering me is that you have port forward code that calls
> remotecommand.NewExecutor just so it can get an httpstream.Dialer that
> you can use to call Dial() so you can get an httpstream.Connection. None
> of that is really about "remote command", at least not in the "exec/attach"
> sense.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/14618/files#r41433747>.
>